fix(46246): Extending Class/Interface With a Primitive Throws a More Concise Error Report

maross3 opened the pull request:
fixes #46246 

### Example
Class:
```TS
class foo extends number {} 
```
Error:
```A class cannot extend a primitive type like 'number'. A class can only extend named object types.```

Interface:
```TS
interface bar extends number { }
```
Error:
```An interface cannot extend a primitive type like 'number'. An interface can only extend named object types.```


### Notes:
* gulp `runtests` passed on current branch.
* I updated a few unit tests to test for the error. I did not feel new test cases were necessary, as the errors were well covered.
* Function responsible for the ambiguous error: `checkAndReportErrorForUsingTypeAsValue`
* Function to check heritage clause: `checkAndReportErrorForExtendingPrimitiveType`

## Description
The approach was to check for extension of primitives directly before the ambiguous error message was thrown. This lead to making a new function.

I didn't add the code to the function `checkAndReportErrorForUsingTypeAsValue` because the intended error checking functionality was separate from evaluating the *value* for error checking. 
 
The new function `checkAndReportErrorForExtendingPrimitiveType` now handles checking heritage clause related errors in regards to classes extending primitives.

In the code, we check if the current errorLocation's expected heritage clause was, in fact, a `SyntaxKind.HeritageClause`. A bitmask was used to check against the expected heritage clause: `errorLocation.parent.parent.kind`. If it's a primitive name and a heritage clause, we throw a more descriptive error.

### This was my first contribution to any project of any size, so I would appreciate any feedback!

MartinJohns commented:
> An interface can only extend named types and classes.

Unless that named type is a union type or a mapped type.

Or as error 2312 frames it:

> An interface can only extend an object type or intersection of object types with statically known members.

DanielRosenwasser commented on tests/baselines/reference/classExtendingPrimitive.errors.txt:
```diff
@@ -14,13 +14,13 @@ tests/cases/conformance/classes/classDeclarations/classHeritageSpecification/cla
     class C3 extends boolean { }
                      ~~~~~~~
-!!! error TS2693: 'boolean' only refers to a type, but is being used as a value here.
+!!! error TS18042: An interface cannot extend a primitive type like 'boolean'; An interface can only extend named types and classes.
```
You're reporting this on a class, not an interface.

maross3 replied:
Changes:
* Two errors to distinguish between class and interface 
* Conditional flow using bitwise operations on `meaning & class` and `meaning & interface`

DanielRosenwasser commented on src/compiler/diagnosticMessages.json:
```diff
@@ -7356,5 +7356,9 @@
         "category": "Error",
         "code": 18041
+    },
+    "An interface cannot extend a primitive type like '{0}'; An interface can only extend named types and classes.": {
```
Don't capitalize after the semicolon.

```suggestion
    "An interface cannot extend a primitive type like '{0}'. An interface can only extend named object types.": {
```

maross3 replied:
Changes:
* Error messages were rewritten to standard

MartinJohns commented:
> Couple quick questions:

No idea. I'm no member of the TypeScript team. I just saw the new error message you added and wanted to point that it's incorrect (or at least not always correct).

maross3 commented:
Thank you,

* Added the logic to distinguish between the interface and class.
* Added two errors, to keep consistent with other functions/error messages.
* Fixed the typos

sandersn commented:
#46357 fixed #46246 earlier, although we dropped it until now. I merged that PR, but this PR also adds an error message on classes, which is pretty useful. If you want to continue on this PR, you could add the class error to the code added by #46357.

sandersn commented on README.md:
````diff
@@ -95,7 +95,7 @@ gulp help              # List the above commands.
 
 ```bash
-node built/local/tsc.js hello.ts
+node <repo-root>/built/local/tsc.js hello.ts
````
can you revert the changes here and in package.json? They're not related to the fix.

maross3 replied:
Sorry, will do.

sandersn commented on src/compiler/diagnosticMessages.json:
```diff
@@ -7356,5 +7356,13 @@
+        "category": "Error",
+        "code": 18042
+    },
+    "A class cannot extend a primitive type like '{0}'. A class can only extend named object types.": {
```
While you're adding class-specific errors, a separate error for when classes *implement* primitive types would be nice.

sandersn commented on src/compiler/checker.ts:
```diff
@@ -2497,6 +2498,22 @@ namespace ts {
+                    else if(meaning & SymbolFlags.Class) {
+                        error(errorLocation, Diagnostics.A_class_cannot_extend_a_primitive_type_like_0_A_class_can_only_extend_named_object_types, rawName);
+                    }
+                     return true;
```
I don't think you should return true unless one of the above errors was issued.

sandersn commented on src/compiler/checker.ts:
```diff
@@ -2497,6 +2498,22 @@ namespace ts {
+                    if(meaning & SymbolFlags.Interface) {
+                        error(errorLocation, Diagnostics.An_interface_cannot_extend_a_primitive_type_like_0_An_interface_can_only_extend_named_object_types, rawName);
+                    }
+                    else if(meaning & SymbolFlags.Class) {
```
A syntactic check (of `errorLocation.parent.parent.parent` I think) is better. `meaning` is the kind of symbol *requested* by name resolution.

Also you should probably check that all those parent nodes are defined.

sandersn commented:
@maross3 Do you want to keep working on this? Otherwise I'd like to close it to reduce the number of open PRs.
